File: tests/omp/simd6.ml
open Optitrust
open Target

let _ = Run.script_cpp (fun _ ->
  !! Omp.declare_simd [Linear (["p"], 1); NotInbranch] [tBefore; cFunDef "foo"];
  !! Omp.simd [] [tBefore; cFunDef "myaddint"; cFor_c "i"];
  !! Omp.declare_simd [Linear (["p"], 1); Inbranch] [tBefore; cFunDef "goo"];
  !! Omp.simd [] [tBefore; cFunDef "myaddfloat"; cFor_c "i"];
  
)

File: tests/omp/simd6.cpp
int foo(int *p){
  *p = *p + 10;
  return *p;
}
 
int myaddint(int *a, int *b, int n)
{
  for (int i=0; i<n; i++){
      a[i]  = foo(&b[i]);  /* foo is not called under a condition */
  }
  return a[n-1];
}
 
float goo(float *p){
  *p = *p + 18.5f;
  return *p;
}
 
int myaddfloat(float *x, float *y, int n)
{
  for (int i=0; i<n; i++){
     x[i] = (x[i] > y[i]) ? goo(&y[i]) : y[i];
       /* goo is called under the condition (or within a branch) */
  }
  return x[n-1];
}